### PR TITLE
[MRG] Laplace regularized OTDA

### PR DESCRIPTION
**Pull request by ievred**
+ added emd with laplacian regularization, test to tests, 
+ laplacian matrix calculation to utils 
+ added example
+ update readme for new POT urml

### CODE REVIEW

File: examples/plot_otda_laplacian.py [hunk 1]
```diff
@@ -0,0 +1,127 @@
+# -*- coding: utf-8 -*-
+"""
+========================
+OT for domain adaptation
```
Comment (rflamary) [inline examples/plot_otda_laplacian.py:4]: add proper titke since it sill appear in the documentation

Comment (ievred) [reply examples/plot_otda_laplacian.py]: done

File: ot/da.py [hunk 2]
```diff
@@ -748,6 +748,115 @@ def OT_mapping_linear(xs, xt, reg=1e-6, ws=None,
...
+    gamma : (ns x nt) ndarray
+        Optimal transportation matrix for the given parameters
+    log : dict
+        log dictionary return only if log==True in parameters
+
+
+    References
+    ----------
+
+    .. [5] N. Courty; R. Flamary; D. Tuia; A. Rakotomamonjy,
```
Comment (rflamary) [inline ot/da.py:817]: please add the paper that provide the expression and gradients here and in the readme. 

You should also update the readme at the line:
Optimal transport for domain adaptation with group lasso regularization [5]

to add Laplacian regularization and also add yourself as a contributor ;)

Comment (ievred) [reply ot/da.py:817]: added Laplacian regularization to readme and workshop citation to da.py in both emd_laplace and EMDLaplace and to readme

Comment (ievred) [reply ot/da.py:817]: and myself to contributors

File: ot/da.py [hunk 3]
```diff
@@ -748,6 +748,118 @@ def OT_mapping_linear(xs, xt, reg=1e-6, ws=None,
...
+
+        sS = kneighbors_graph(xs, kwargs['nn']).toarray()
+        sS = (sS + sS.T) / 2
+        sT = kneighbors_graph(xt, kwargs['nn']).toarray()
+        sT = (sT + sT.T) / 2
+
+    lS = laplacian(sS)
+    lT = laplacian(sT)
+
+    def f(G):
```
Comment (rflamary) [inline ot/da.py:851]: could you please implement both position regularization and displacement regularization?

It only change the M matrix and provides more generality to teh function?

See equation (11) vs (9)
https://remi.flamary.com/biblio/flamary2014optlaplace.pdf

Comment (ievred) [reply ot/da.py:855]: done

File: ot/da.py [hunk 4]
```diff
@@ -748,6 +748,118 @@ def OT_mapping_linear(xs, xt, reg=1e-6, ws=None,
...
+        return alpha * np.trace(np.dot(xt.T, np.dot(G.T, np.dot(lS, np.dot(G, xt))))) \
+            + (1 - alpha) * np.trace(np.dot(xs.T, np.dot(G, np.dot(lT, np.dot(G.T, xs)))))
+
+    def df(G):
+        return alpha * np.dot(lS + lS.T, np.dot(G, np.dot(xt, xt.T)))\
```
[the remark this answers is not in the document]
Comment (ievred) [reply ot/da.py]: done

File: ot/da.py [hunk 5]
```diff
@@ -748,6 +748,118 @@ def OT_mapping_linear(xs, xt, reg=1e-6, ws=None,
...
+    elif sim == 'knn':
+        if 'nn' not in kwargs:
+            kwargs['nn'] = 5
+
+        from sklearn.neighbors import kneighbors_graph
+
+        sS = kneighbors_graph(xs, kwargs['nn']).toarray()
+        sS = (sS + sS.T) / 2
+        sT = kneighbors_graph(xt, kwargs['nn']).toarray()
+        sT = (sT + sT.T) / 2
```
[the remark this answers is not in the document]
Comment (ievred) [reply ot/da.py:848]: done

File: ot/da.py [hunk 6]
```diff
@@ -1576,6 +1701,120 @@ class label
...
+    ----------
+    .. [1] N. Courty; R. Flamary; D. Tuia; A. Rakotomamonjy,
+           "Optimal Transport for Domain Adaptation," in IEEE Transactions
+           on Pattern Analysis and Machine Intelligence , vol.PP, no.99, pp.1-1
+    .. [2] R. Flamary, N. Courty, D. Tuia, A. Rakotomamonjy,
+        "Optimal transport with Laplacian regularization: Applications to domain adaptation and shape matching,"
+         in NIPS Workshop on Optimal Transport and Machine Learning OTML, 2014.
+    """
+
+    def __init__(self, reg_type='pos', reg_lap=1., reg_src=1., alpha=0.5,
```
Comment (rflamary) [inline ot/da.py:1756]: how can you pass nn or sim to the emd_laplace function?
this is necessary, you should use a kwargs and store it for the fit

File: ot/da.py [hunk 7]
```diff
@@ -748,6 +748,131 @@ def OT_mapping_linear(xs, xt, reg=1e-6, ws=None,
...
+    Parameters
+    ----------
+    a : np.ndarray (ns,)
+        samples weights in the source domain
+    b : np.ndarray (nt,)
+        samples weights in the target domain
+    xs : np.ndarray (ns,d)
+        samples in the source domain
+    xt : np.ndarray (nt,d)
+        samples in the target domain
```
Comment (rflamary) [inline ot/da.py:785]: sim not documented. You should add it to the list and define also the parameters for each sim : 'nn' and the gaussian kernel width.

File: ot/da.py [hunk 8]
```diff
@@ -748,6 +748,131 @@ def OT_mapping_linear(xs, xt, reg=1e-6, ws=None,
...
+    See Also
+    --------
+    ot.lp.emd : Unregularized OT
+    ot.optim.cg : General regularized OT
+
+    """
+    if sim == 'gauss':
+        if 'rbfparam' not in kwargs:
+            kwargs['rbfparam'] = 1 / (2 * (np.mean(dist(xs, xs, 'sqeuclidean')) ** 2))
+        sS = kernel(xs, xs, method=kwargs['sim'], sigma=kwargs['rbfparam'])
```
Comment (rflamary) [inline ot/da.py:836]: do not use kwargs['sim'], use 'gauss'

File: ot/da.py (hunk 6 above)
Comment (ievred) [reply ot/da.py:1767]: added dict kwargs in fit that is passed to emd_laplace

File: ot/da.py (hunk 7 above)
Comment (ievred) [reply ot/da.py:785]: documented sim and added a common parameter 'similarity_param'

File: ot/da.py (hunk 8 above)
Comment (ievred) [reply ot/da.py]: I left it in case one day another kernel wille be added to kernel but can fix it
